Estudiante: Valeria Esperanza Bravo Saines

Instalación de Pycaret:

Pipeline de pycaret:

1. Setup(): Definir el dataset y parámetros

2. Compare_models(): Comparar multiples modelos y los rankear

3. Create model(): Crear un modelo específico

4. Tune_model(): Ajuste automático de hiperparámetros

5. plot_model(): visualizar el desempeño del modelo

6. Evaluate_model(): Evaluación del modelo

7. Predict_model(): Predicciones sobre nuevos datos en el DF

8. Save_model/load_model(): Guardar y cargar el modelo

In [9]:
import pycaret
print(pycaret.__version__)

3.3.2


In [10]:
import pandas as pd
from pycaret.classification import setup, compare_models, create_model, evaluate_model, save_model
from pycaret.classification import pull
from pycaret.classification import *


In [ ]:
df = pd.read_csv('ds_credito_vehicular.csv')

In [12]:
import os
print(os.getcwd())

c:\Users\vbravo\Downloads\CLASE 5 -MLE\Machine-Learning-Engineering\Material sesión 05-ML E


In [13]:
df.head()

edad  ingresos_mensuales estado_civil historia_credito  \
0    59                7668   Divorciado            Buena   
1    49                6279       Casado            Buena   
2    35                2722      Soltero             Mala   
3    63                4314   Divorciado          Regular   
4    28                4157      Soltero            Buena   

   nro_creditos_previos  cuota_vs_ingreso  vehiculo_propio  default  
0                     2              0.64                0        1  
1                     2              0.31                1        1  
2                     4              0.41                1        0  
3                     4              0.57                0        0  
4                     1              0.34                0        0

In [14]:
#1. Setup() — definir el dataset y los parámetros
red = setup(
    data=df, target="default", session_id=124, normalize=True, categorical_features=["vehiculo_propio", "historia_credito", "estado_civil"], fix_imbalance=True
)

In [15]:
#2. Compare_models(): Comparar multiples modelos y los rankear
best_model = compare_models()

Initiated  . . . . . . . . . . . . . . . . . .              17:37:54
Status     . . . . . . . . . . . . . . . . . .  Loading Dependencies
Estimator  . . . . . . . . . . . . . . . . . .     Compiling Library

In [16]:
#guardamos los resulatdo en un excel
results = pull()
results.to_csv('resultados_modelos.csv', index=False, sep=';')


EMPEZAMOS CON EL MODELO ESCOGIDO

In [17]:
#3. Create model(): Crear un modelo específico
lr = create_model('lr')


In [18]:
print(best_model)
#El mejor modelo fue la regresion logistica segun la métrica AUC

DummyClassifier(constant=None, random_state=124, strategy='prior')


In [ ]:
#5. plot_model(): visualizar el desempeño del modelo
#6. Evaluate_model(): Evaluación del modelo
evaluate_model(best_model)


interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [21]:
#7. Predict_model(): Predicciones sobre nuevos datos en el DF
preds = predict_model(best_model)
print(preds.head())

     edad  ingresos_mensuales estado_civil historia_credito  \
75     43                2707   Divorciado            Buena   
47     22                2495   Divorciado            Buena   
1      49                6279       Casado            Buena   
92     64                2930       Casado            Buena   
135    46                1623       Casado            Buena   

     nro_creditos_previos  cuota_vs_ingreso  vehiculo_propio  default  \
75                      4              0.30                1        0   
47                      3              0.65                1        0   
1                       2              0.31                1        1   
92                      2              0.31                1        0   
135                     2              0.18                1        0   

     prediction_label  prediction_score  
75                  0               0.5  
47                  0               0.5  
1                   0               0.5  
92            

In [20]:
#8. Save_model/load_model(): Guardar y cargar el modelo
save_model(best_model, 'modelo_credito_final')


Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['edad', 'ingresos_mensuales',
                                              'nro_creditos_previos',
                                              'cuota_vs_ingreso'],
                                     transformer=SimpleImputer(add_indicator=False,
                                                               copy=True,
                                                               fill_value=None,
                                                               keep_empty_features=False,
                                                               missing_values=nan,
                                                               strategy='mean'))),
                 ('categorical_imputer',
                  TransformerWr...
                  TransformerWrapper(exclude=None, include=None,
                      

CONCLUSIONES 1:

1. Debido a que variable defualt poseia solo dos valores, si o no, se uso un modelo de clasificación binaria.

2. El modelo con mayor desempeño fue la regresion logistica, donde su resultado mostró un AUC medio, lo que indica un rendimiento aceptable.

3. Se probó el modelo de regresión logistica para predecir si un cliente incumplirA su credito.

4. El modelo obtuvo un desempeño moderado segun las metricas de AUC y F1, y predice correctamente la mayoría de los casos negativos, pero falla en algunos positivos.

5. No se aplico ajuste de parametros (tuning), por lo que aún puede mejorar

MEJORA DEL MODELO

In [22]:
#1. Setup() — definir el dataset y los parámetros
clf = setup(
 data=df, target='default', session_id=123, normalize=True, categorical_features=['vehiculo_propio', 'historia_credito', 'estado_civil'],
 fix_imbalance=True
)

In [25]:
#2. Compare_models(): Comparar multiples modelos y los rankear
best_model = compare_models()

In [24]:
# 3. El modelo ya se creó 
#4. Tune_model(): Ajuste automático de hiperparámetros
tuned_model = tune_model(best_model, optimize='F1')

Fitting 10 folds for each of 4 candidates, totalling 40 fits


In [ ]:
#5. plot_model(): visualizar el desempeño del modelo
#6. Evaluate_model(): Evaluación del modelo
evaluate_model(tuned_model)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [ ]:
#7. Predict_model(): Predicciones sobre nuevos datos en el DF
preds_tuned = predict_model(tuned_model)
print(preds_tuned.head())

In [ ]:
#8. Save_model/load_model(): Guardar y cargar el modelo
save_model(best_model, 'modelo_credito_tuned_final')


CONCLUSIONES 2:

1. Antes del tuning, la regresión logistica tuvo un rendimiento regular, con una exactitud de 0.57 más o menos.

2. Al aplicar el tuning, subió un poco a 0.61, asi que mejoro ligeramente.

3. El AUC sigue en 0.50, o sea que el modelo todavía no logra diferenciar bien las clases, tambien el F1 bajó un poco, así que el ajuste no ayudo mucho en el balance entre precisiOn y recall.

4. De forma general tuvo una mejora minima, se podría probar con otros modelos pero por el momento el adeucado es el escogido.